### PR TITLE
C synth

### PR DESCRIPTION
**Pull request by nwdepatie**
## Changes

This is my first go at creating a C Synthesizer using our cangen tooling. After looking at all of the tooling inside of cangen, I realized that it was super super tailored towards rust meaning I needed to go in and change a lot of stuff to get this working. Later on, I also realized that the CAN config felt really odd and some of the data structures we felt a bit weird too, so I might clean some more stuff up. Keep in mind I was making this on trains and planes lmao

## Test Cases

- Running locally with a test script to verify output
- Verifying that Calypso still builds with my changes

## To Do

_Any remaining things that need to get done_

- [x] Test C compilation
- [x] Improve the structures that define a CAN message a bit
- [x] Make Python code and data structures more generalized

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No merge conflicts
- [x] All checks passing
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: cangen/CANField.py [hunk 1]
```diff
@@ -7,13 +7,46 @@
...
+            num_bytes += 1
+
+        return int(num_bytes)
+
+@dataclass
+class DiscreteField(CANField):
+    """
+    Represents a discrete field inside a CAN message in the sense that
+    it will make sense if it is published to MQTT and timestamped on its own
+    (think State of Charge or Temperature)
```
Comment (Peyton-McKee) [inline cangen/CANField.py:38]: (Single Point of Data)

File: cangen/CANField.py [hunk 2]
```diff
@@ -7,13 +7,46 @@
...
+
+@dataclass
+class CompositeField(CANField):
+    """
+    Represents a composite field inside a CAN message in the sense that
+    it will only make sense with other data points in the same message and
+    can't be timestamped on its own (think XYZ acceleration or Roll, Pitch, Yaw)
+    """
+
+    num_points: int = 0
```
Comment (Peyton-McKee) [inline cangen/CANField.py:51]: dont set a default

Comment (Peyton-McKee) [reply cangen/CANField.py:51]: (means you have to put it under field type)

File: cangen/CANMsg.py [hunk 3]
```diff
@@ -1,22 +1,25 @@
...
+    desc: str   # Brief name of CAN message, used for generating function names
+    fields: list[CANField] # List of CAN fields in the message (Optional to allow for only networkEncoding)
 
     def __post_init__(self) -> None:
         idx: int = 0
-        for field in self.networkEncoding[0].fields:
-            if (field.index is not None):
-                field.index = idx
-                idx += field.size
+
```
Comment (Peyton-McKee) [inline cangen/CANMsg.py:17]: just remove this entire thing

File: cangen/CANMsg.py [hunk 4]
```diff
@@ -1,22 +1,25 @@
...
 class CANMsg:
     '''
     Represents a CAN message. Has an id, a description, and a number of individual fields.
     '''
-    id: str
-    desc: str
-    networkEncoding: list[NetworkEncoding]
+    id: str     # Hex value of CAN ID, i.e. `0x88`
+    desc: str   # Brief name of CAN message, used for generating function names
+    fields: list[CANField] # List of CAN fields in the message (Optional to allow for only networkEncoding)
```
Comment (Peyton-McKee) [inline cangen/CANMsg.py:13]: not optional anymore right

File: cangen/CSynth.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
+from cangen.CANField import CANField
+from cangen.CANMsg import CANMsg
+from cangen.Result import Result
+from cangen.Decoding import Decoding
+from typing import List
+from dataclasses import dataclass
+
+""" @dataclass
```
Comment (Peyton-McKee) [inline cangen/CSynth.py]: do we ned these here?

File: cangen/CSynth.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    repr: str = "twos_comp"
+    entry_type = "uint16_t"
+
+@dataclass
+class Half(Decoding):
+    repr: str = "half"
+    entry_type = "uint8_t" """
+
+"""
+Example prototype
```
Comment (Peyton-McKee) [inline cangen/CSynth.py:8]: should we simulataneously generate code for sending and receiving

File: cangen/CSynth.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+}
+"""
+
+class CSynth:
+	"""
+	A class to synthesize C from a given CANMsg spec.
+	"""
+
+	decode_data_import: str = (
+		'#include <stdio.h>\n#include <stdint.h>\n#include "can.h"\n\n'  # Including the necessary headers
```
Comment (Peyton-McKee) [inline cangen/CSynth.py]: can we use a text block for clarity

File: cangen/CSynth.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+		return result
+
+	# Helper function that synthesizes the decode function for a given CANMsg
+	def synthesize(self, msg: CANMsg) -> str:
+		signature: str = self.signature(msg.desc, msg.networkEncoding[0].fields)
+		generated_lines: list[str] = []
+
+		# Generate a line for each field in the message
+		generated_lines += self.parse_network_encoding(msg)
```
Comment (Peyton-McKee) [inline cangen/CSynth.py:51]: i thought we got rid of network encodings?

File: cangen/CSynth.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+		return "\n".join(total_list)
+
+	def format_msg (self, msg: CANMsg) -> str:
+		return f"    can_msg_t mc_msg;\n    mc_msg.id = {msg.id};\n    mc_msg.len = sizeof(mc_msg_data);"
+
+	def parse_network_encoding(
+		self, msg: CANMsg
+	) -> list[str]:  # Change return type to list[str]
+		result = []
+		networkEncoding = msg.networkEncoding[0]
```
Comment (Peyton-McKee) [inline cangen/CSynth.py:68]: this doesnt exist anymore on a canmsg

File: cangen/CSynth.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+			[signature, self.packed_struct_value]
+			+ generated_lines
+			+ [self.packed_struct_end]
+			+ [self.format_msg(msg)]
+			+ [self.decode_close]
+		)
+		return "\n".join(total_list)
+
+	def format_msg (self, msg: CANMsg) -> str:
+		return f"    can_msg_t mc_msg;\n    mc_msg.id = {msg.id};\n    mc_msg.len = sizeof(mc_msg_data);"
```
Comment (Peyton-McKee) [inline cangen/CSynth.py:62]: im in favor of text blocks

File: cangen/RustSynth.py [hunk 11]
```diff
@@ -1,120 +1,97 @@
 from cangen.CANField import CANField
 from cangen.CANMsg import CANMsg
 from cangen.Result import Result
+from cangen.Decoding import BigEndian, LittleEndian, TwosComplement
 from typing import List
 
+TwosComplement.entry_type = "u32"
```
Comment (Peyton-McKee) [inline cangen/RustSynth.py:7]: why is this here

File: cangen/RustSynth.py [hunk 12]
```diff
@@ -1,120 +1,97 @@
...
+        """
+        Helper function that synthesizes the decode function for a given CANMsg
+        """
+
+        # Generate Function Signature
         signature: str = self.signature(msg.desc)
-        length_check: str = self.add_length_check(msg.networkEncoding[0].fields)
-        generated_lines: list[str] = []
+
+        # TODO: I think we don't really need length check
```
Comment (Peyton-McKee) [inline cangen/RustSynth.py:52]: need length check for defensive coding

File: cangen/RustSynth.py [hunk 13]
```diff
@@ -1,120 +1,97 @@
...
-        elif networkEncoding.id == "single_point":
-            for field in networkEncoding.fields:
-                result.append(f"        {networkEncoding.start}")
+
+        # For all the fields in the CAN message, generate the required code
+        for field in msg.fields:
+
+            # If the field is discrete
+            if field.field_type == "discrete":
+                result.append(f"        {RustSnippets.network_encoding_start}")
```
Comment (Peyton-McKee) [inline cangen/RustSynth.py:76]: this can go outside of the if statement its the same for both

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
This looks good. Im glad that the synth looks pretty much the same for rust but the yaml looks much cleaner and hopefully should be more customizable for us in the future.

File: cangen/CANField.py (hunk 2 above)
Comment (nwdepatie) [reply cangen/CANField.py:50]: Sadly Python yells at me if I don't default it

File: cangen/CANField.py [hunk 14]
```diff
@@ -7,13 +7,46 @@
...
     index: int = -1
     final_type: str = "f32"
-    decodings: Optional[list[Decoding]] = None
+    encodings: Optional[list[Decoding]] = None
     format: Optional[str] = None
+
+    def get_size_bytes(self):
+        # Calculate max number of bytes to represent value
+        num_bytes = self.size / 8
+        if (self.size % 8):
```
[the remark this answers is not in the document]
Comment (nwdepatie) [reply cangen/CANField.py:27]: I could but then I'd need to take the ceiling for like different multiples of 8

File: cangen/CANField.py [hunk 15]
```diff
@@ -7,13 +7,46 @@
...
     index within its parent CANMsg, which is assigned at load from YAML.
+
+    Think 1 MQTT Topic per 1 CAN Field
     '''
+
     name: str
     unit: str
-    size: int
+    size: int # in bits
+    field_type: str = "*"*42
```
[the remark this answers is not in the document]
Comment (nwdepatie) [reply cangen/CANField.py:20]: Kinda just a thing to default the base class to something, I think Charlie did something similar somewhere else in the code and I vibed with it. Kinda just like "***************************"

**Comment by nwdepatie** [comment]
Closing as I want to make a larger PR for this and maybe merge in smaller branches in. New PR can be found here on a new develop branch https://github.com/Northeastern-Electric-Racing/Embedded-Base/pull/91
